dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -409,113 +406,60 @@ private RefinementIteration doRefinementIteration(ProverEnvironment prover, WMMS
-        }
-    }
-
-    private static Relation getCopyOfRelation(Relation rel, Wmm m) {
-        Optional<String> name = rel.getName();
-        if (name.isPresent() && m.containsRelation(name.get())) {
-            return m.getRelation(name.get());
-        }
-        Relation copy = name.map(m::newRelation).orElseGet(m::newRelation);
-        return m.addDefinition(rel.getDefinition().accept(new RelationCopier(m, copy)));
+        return RefinementModel.fromCut(Cut.computeInducedCut(original, constraintsToCut));
     }
 
-    private static final class RelationCopier implements Definition.Visitor<Definition> {
-        final Wmm targetModel;
-        final Relation relation;
-
-        RelationCopier(Wmm m, Relation r) {
-            targetModel = m;
-            relation = r;
-        }
-
-        @Override public Definition visitUnion(Relation r, Relation... o) { return new Union(relation, copy(o)); }
-        @Override public Definition visitIntersection(Relation r, Relation... o) { return new Intersection(relation, copy(o)); }
-        @Override public Definition visitDifference(Relation r, Relation r1, Relation r2) { return new Difference(relation, copy(r1), copy(r2)); }
-        @Override public Definition visitComposition(Relation r, Relation r1, Relation r2) { return new Composition(relation, copy(r1), copy(r2)); }
-        @Override public Definition visitInverse(Relation r, Relation r1) { return new Inverse(relation, copy(r1)); }
-        @Override public Definition visitDomainIdentity(Relation r, Relation r1) { return new DomainIdentity(relation, copy(r1)); }
-        @Override public Definition visitRangeIdentity(Relation r, Relation r1) { return new RangeIdentity(relation, copy(r1)); }
-        @Override public Definition visitTransitiveClosure(Relation r, Relation r1) { return new TransitiveClosure(relation, copy(r1)); }
-        @Override public Definition visitIdentity(Relation r, Filter filter) { return new Identity(relation, filter); }
-        @Override public Definition visitProduct(Relation r, Filter f1, Filter f2) { return new CartesianProduct(relation, f1, f2); }
-        @Override public Definition visitFences(Relation r, Filter type) { return new Fences(relation, type); }
-
-        private Relation copy(Relation r) { return getCopyOfRelation(r, targetModel); }
-        private Relation[] copy(Relation[] r) {
-            Relation[] a = new Relation[r.length];
-            for (int i = 0; i < r.length; i++) {
-                a[i] = copy(r[i]);
-            }
-            return a;
-        }
-    }
-
-    private Wmm createDefaultWmm() {
-        Wmm baseline = new Wmm();
-        Relation rf = baseline.getRelation(RF);
-        if (baselines.contains(Baseline.UNIPROC)) {
+    private static void addBiases(Wmm memoryModel, EnumSet<Baseline> biases) {
+        // FIXME: This can (in theory) add redundant intermediate relations and/or constraints that
+        //  already exist in the model.
```

Comment:
It should just add redundancy that can affect performance. This issue is not specific to biases but is more general with how we handle predefined relations/axioms.
IIRC, this issue already appears in `RISCV`.
The issue is in code like this:
```
let r1 = fr | ... // Implicitily defines a "rf^-1" relation
let r2 = rf^-1 // Explicitly defines another "rf^-1"
```
For biases, it is no practical issue because the baseline model has almost no relations to begin with (so there is hardly a possiblity to create redundancy), unless we cut a lot relations from the original model.